concurrent/src/main/java/io/airlift/concurrent/AsyncSemaphore.java

```diff
@@ -37,26 +42,70 @@
  * to do the bulk of its work asynchronously.
  */
 @ThreadSafe
-public class AsyncSemaphore<T>
+public class AsyncSemaphore<T, R>
 {
-    private final Queue<QueuedTask<T>> queuedTasks = new ConcurrentLinkedQueue<>();
+    private final Queue<QueuedTask<T, R>> queuedTasks = new ConcurrentLinkedQueue<>();
     private final AtomicInteger counter = new AtomicInteger();
     private final Runnable runNextTask = this::runNext;
     private final int maxPermits;
     private final Executor submitExecutor;
-    private final Function<T, ListenableFuture<?>> submitter;
+    private final Function<T, ListenableFuture<R>> submitter;
+
+    /**
+     * Process a list of tasks as a single unit
+     * (similar to {@link com.google.common.util.concurrent.Futures#allAsList(ListenableFuture[])})
+     * with limiting the number of tasks running in parallel.
+     * <p>
+     * This method may be useful for limiting the number of concurrent requests sent to a remote server when
+     * trying to load multiple related entities concurrently:
+     * <p>
+     * For example:
+     * <pre>{@code
+     * List<Integer> userIds = Lists.of(1, 2, 3);
+     * ListenableFuture<List<UserInfo>> future = processAll(ids, client::getUserInfoById, 2, executor);
+     * List<UserInfo> userInfos = future.get(...);
+     * }</pre>
+     *
+     * @param tasks tasks to process
+     * @param submitter task submitter
+     * @param maxPermits maximum number of tasks permitted to run in parallel
+     * @param executor task submission executor
+     * @return {@link ListenableFuture} containing a list of values returned by the {@code tasks}.
+     * The order of elements in the list matches the order of {@code tasks}.
+     * If the result future is cancelled all the remaining tasks are cancelled (submitted tasks will be cancelled, pending tasks will not be submitted).
+     * If any of the submitted tasks fails or is canceled, the result future is too.
+     * If any of the submitted tasks fails or is canceled, the remaining tasks are cancelled.
```

Comment:
"submitted tasks fails or is canceled" => "submitted tasks fail or are cancelled"